certora/specs/GuardianStorage.spec

```diff
@@ -0,0 +1,378 @@
+    }
+
+// every element reaches the 0 pointer (because we replace in reach the end sentinel with null)
+invariant reach_null()
+    (forall address wallet. forall address X. reach(wallet, X, NULL))
+    {
+        preserved {
+            requireInvariant reach_invariant();
+            requireInvariant inListReachable();
+            requireInvariant reachableInList();
+        }
+    }
+
+// every element that is reachable from another element is either the null pointer or part of the list.
+invariant reachableInList()
+    (forall address wallet. forall address X. forall address Y. reach(wallet, X, Y) => X == Y || Y == 0 || ghostGuardians[wallet][Y] != 0)
+    {
+        preserved {
+            requireInvariant reach_invariant();
+            requireInvariant reach_null();
+            requireInvariant inListReachable();
+            requireInvariant reach_next();
+            requireInvariant nextNull();
+            requireInvariant countZeroIffListEmpty();
+        }
+    }
+
+// reach encodes a linear order. This axiom corresponds to Table 2 in [1].
+invariant reach_invariant()
+    forall address wallet. forall address X. forall address Y. forall address Z. (
+        reach(wallet, X, X)
+        && (reach(wallet, X, Y) && reach(wallet, Y, X) => X == Y)
+        && (reach(wallet, X, Y) && reach(wallet, Y, Z) => reach(wallet, X, Z))
+        && (reach(wallet, X, Y) && reach(wallet, X, Z) => (reach(wallet, Y, Z) || reach(wallet, Z, Y)))
+    )
+    {
+        preserved {
+            requireInvariant reach_null();
+            requireInvariant inListReachable();
+            requireInvariant reachableInList();
+            requireInvariant reachHeadNext();
+        }
+    }
+
+// every element with non-zero guardian field is reachable from SENTINEL (head of the list)
+invariant inListReachable()
+    (forall address wallet. forall address key. ghostGuardians[wallet][key] != 0 => reach(wallet, SENTINEL, key))
+    {
+        preserved {
+            requireInvariant reach_invariant();
+            requireInvariant reach_null();
+            requireInvariant reachableInList();
+            requireInvariant countZeroIffListEmpty();
+            requireInvariant emptyListNotReachable();
+        }
+    }
+
+invariant reachHeadNext()
+    forall address wallet. forall address X. (reach(wallet, SENTINEL, X) && X != SENTINEL && X != NULL) =>
+           (ghostGuardians[wallet][SENTINEL] != SENTINEL && reach(wallet, ghostGuardians[wallet][SENTINEL], X))
+    {   
+        preserved {
+            requireInvariant inListReachable();
+            requireInvariant reachableInList();
+            requireInvariant reach_invariant();
+            requireInvariant reach_null();
+            requireInvariant countZeroIffListEmpty();
+            requireInvariant emptyListNotReachable();
+            requireInvariant nextNull();
+        }
+    }
+
+// every element reaches its direct successor (except for the tail-SENTINEL).
+invariant reach_next()
```

Comment:
nit: the naming is inconsistent between snake case and camel case

Reply:
https://github.com/5afe/CandideWalletContracts/pull/20/commits/35684c169cf8c8974aa5291367dca7b17b0161bd